### PR TITLE
Codes for collecting diagnostic from Twelf server and send it back to vscode

### PR DESCRIPTION
**Pull request by UltimatePea**
I've used unopensourced code (possibly violating the copyright of the original author) from sml language extensions, but this should work as basic Twelf checking in VSCode.

### CODE REVIEW

File: extension.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	let restartRepl = vscode.commands.registerCommand('twelf-server.restart', () => smlEnviron.restart());
+	// let execCurrentFile = vscode.commands.registerCommand('sml-environment.execCurrentFile', () => smlEnviron.execCurrentFile())
+	
+	// context.subscriptions.push(execShortCode);
+	context.subscriptions.push(restartRepl);	
+	// context.subscriptions.push(execCurrentFile)
+
+	context.subscriptions.push(vscode.workspace.onDidSaveTextDocument(smlEnviron.didSaveDocument));
+}
+exports.activate = activate;
```
Comment (yaene) [inline extension.js]: I think this line can just be removed since exports are overwritten again at the end of this file

File: extension2.js [hunk 2]
```diff
@@ -0,0 +1,281 @@
+// COPIED FROM https://marketplace.visualstudio.com/items?itemName=freebroccolo.sml
```
Comment (yaene) [inline extension2.js]: extension2.js is a bit of a curious name.
I also couldn't find any references to this file anywhere, what is it for?

File: package.json [hunk 3]
```diff
@@ -33,6 +37,20 @@
...
+          "type": "string",
+          "title": "Twelf interpreter path",
+          "default": "/usr/local/bin/twelf-server",
+          "comment": "defaults to /usr/local/bin/twelf-server"
+        }
+      }
+    }
+  },
+  "dependencies": {
+    "file-uri-to-path": "^2.0.0"
```
Comment (yaene) [inline package.json]: The only place i found where this dependency is required is commented out. It can probably be removed again.

**Review by yaene — COMMENTED** [review]
Thanks for the contribution!
I didn't get to test it yet but just had a quick look at the code.
Review points:
- lots of commented code that could probably be removed
- see comments made directly in code

File: extension2.js (hunk 2 above)
Comment (UltimatePea) [reply extension2.js]: It is used for retaining copyright to the original author (whose code is not public), 

I have moved this notice to the main file and deleted the files

File: package.json (hunk 3 above)
Comment (UltimatePea) [reply package.json]: removed

File: extension.js (hunk 1 above)
Comment (UltimatePea) [reply extension.js]: okay

**Comment by UltimatePea** [comment]
Hi there,

I've made changes according to your suggestions. But again, a part of the code  is copied from https://marketplace.visualstudio.com/items?itemName=freebroccolo.sml, whose author (@freebroccolo) did not choose to make it public. So merge with care.

We should probably rewrite those parts if the copyright holder complains. Otherwise, I think it is perfectly fine for personal use.

**Comment by UltimatePea** [comment]
The problematic codes are in smlEnvironmentManager.js line 24 - 53

**Comment by yaene** [comment]
Ah I see.
Btw how did you find out it is not open source?
I checked the license that's linked on the market place and its apparently apache, so we should be fine as long as we clearly mention which part is under apache license and link the license?
https://marketplace.visualstudio.com/items/freebroccolo.sml/license

**Comment by UltimatePea** [comment]
Given they used Apache, I guess we're good to go then. 

I basically took the codes from both 
[https://marketplace.visualstudio.com/items/freebroccolo.sml](https://marketplace.visualstudio.com/items/freebroccolo.sml)
and 
https://marketplace.visualstudio.com/items?itemName=vrjuliao.sml-environment, 
adapting them to twelf.


We perhaps need to acknowledge those projects.

The codes can be extracted from the extension marketplace since it's all javascript. You can "unzip" the downloaded vsix file from the vscode marketplace and look at the code by changing ".vsix" to ".zip". It's not open source per se because the associated GitHub repo is private.

**Comment by yaene** [comment]
okay thanks for the info!
I'm sorry for replying a bit slowly, exams are  around the corner. 
Since you got the code from the place where I also found the open source license it should be fine. 
I should be able to properly check out the new things you added next week.

File: extension.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * @param {vscode.ExtensionContext} context
+ */
+function activate(context) {
+	// start sml exection
+	smlEnviron.start();
+	console.log('Congratulations, your extension "twelf" is now active!');
+	
+	// let execShortCode = vscode.commands.registerCommand('sml-environment.execShortCode', () => smlEnviron.execShortCode());
+	let restartRepl = vscode.commands.registerCommand('twelf-server.restart', () => smlEnviron.restart());
```
Comment (yaene) [inline extension.js:18]: Should this command be exposed to the user in the package.json contributes section? Right now i dont see where this command is called and I cant find it in the command palette.

**Review by yaene — COMMENTED** [review]
I tested it and its quite useful!
I only made a small comment so you can check that and after that I'll merge.

File: extension.js (hunk 4 above)
Comment (UltimatePea) [reply extension.js:18]: Fixed now

**Pull request merged**